Ambari Views Integration

rashidaligee opened the pull request:
This code lays the foundation for ambari views. Current commit installs Files views but others can be added by adding new resources/recipes.

cbaenziger commented on cookbooks/ambari/LICENSE:
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
We have the ASF license for the entire repo at the root; I do not believe there is a need for duplication below.

cbaenziger commented on cookbooks/ambari/metadata.rb:
```diff
@@ -0,0 +1,9 @@
+name 'ambari'
+maintainer 'Artem Ervits'
```
Though @dbist did a nice job with the initial work for the Ambari cookbook please use:
```
maintainer       'Bloomberg Finance L.P.'
maintainer_email 'hadoop@bloomberg.net'
license 'Apache License 2.0'
```

cbaenziger commented on cookbooks/ambari/metadata.rb:
```diff
@@ -0,0 +1,9 @@
+long_description 'Installs/Configures ambari-views-chef'
+version '0.1.0'
+
+depends 'postgresql', '~> 6.1.1'
```
Do we anticipate newer version to not work with this? If not, it is best to float with upstream so we know breakages sooner rather than later.

cbaenziger commented on cookbooks/ambari/recipes/ambari_server_setup.rb:
```diff
@@ -0,0 +1,37 @@
+
+
+# /etc/ambari-server/conf/ambari.properties
+template "#{node['ambari']['ambari_server_conf_dir']}ambari.properties" do
```
Please use [`File.join`](http://ruby-doc.org/core-2.2.0/File.html#method-c-join) so if one does not leave a trailing `/` on their directory things will not mysteriously fail.

cbaenziger commented on cookbooks/ambari/recipes/ambari_server_setup.rb:
```diff
@@ -0,0 +1,37 @@
+end
+
+# /etc/ambari-server/conf/password.dat
+template "#{node['ambari']['ambari_server_conf_dir']}password.dat" do
```
Again with `File.join`

cbaenziger commented on cookbooks/ambari/recipes/ambari_server_setup.rb:
```diff
@@ -0,0 +1,37 @@
+
+# /etc/ambari-server/conf/krb5JAASLogin.conf
+if node['ambari']['kerberos']['enabled']
+  template "#{node['ambari']['ambari_server_conf_dir']}krb5JAASLogin.conf" do
```
Again with `File.join`

cbaenziger commented on cookbooks/ambari/recipes/ambari_server_setup.rb:
```diff
@@ -0,0 +1,37 @@
+
+# /etc/ambari-server/conf/password.dat
+template "#{node['ambari']['ambari_server_conf_dir']}password.dat" do
+  source 'password.dat.erb'
```
This looks like a password, please use [`sensitive`](http://blog.thefourthparty.com/chef-sensitive/)

cbaenziger commented on cookbooks/ambari/recipes/ambari_views_setup.rb:
```diff
@@ -0,0 +1,40 @@
+
+http_request 'Polling Ambari WebUi' do
+  url "#{node['ambari']['ambari_server_base_url']}"
+  retries 24
```
Any reason to do 24 retries? It seems like 120 seconds is a long time to wait...

rashidaligee replied:
I just chose a random number. Usually, ambari-server comes up within few seconds but I set the limit to 2 minutes just to give ample time to start ambari server and its components.

cbaenziger commented on cookbooks/ambari/recipes/ambari_views_setup.rb:
```diff
@@ -0,0 +1,40 @@
+end
+
+# create an instance of Files View
+bash 'create_instance_of_Files_view' do
```
Is this REST call idempotent?

rashidaligee replied:
This is kinda idempotent as the first call create the resource but later calls are rejected as the resource is already created.

cbaenziger commented on cookbooks/ambari/recipes/default.rb:
```diff
@@ -0,0 +1,28 @@
+      components ['main']
+      distribution 'Ambari'
+      action :add
+      keyserver 'keyserver.ubuntu.com'
```
Please allow `keyserver` and `key` to be optionally passed in; then we can take advantage of the dual-nature of the `key` attribute:
```If a keyserver is provided, this is assumed to be the fingerprint; otherwise it can be either the URI to the GPG key for the repo, or a cookbook_file. Default value: nil.``` - From the [Chef 12.15 Docs](https://github.com/chef/chef-web-docs/blob/2cdc74c9bbe78db723758715f60ba0b5eb46d394/chef_master/source/resource_apt_repository.rst)

cbaenziger commented on cookbooks/ambari/test/integration/helpers/serverspec/spec_helper.rb:
```diff
@@ -0,0 +1,8 @@
+require 'serverspec'
+
+if (/cygwin|mswin|mingw|bccwin|wince|emx/ =~ RUBY_PLATFORM).nil?
```
Do we expect this to be run on Windows?

rashidaligee replied:
No, I'm not expecting this to run on a windows system. This must have been auto-created. I can remove it if it causes any confusion.

cbaenziger commented on cookbooks/bach_ambari/Berksfile:
```diff
@@ -0,0 +1,17 @@
+cookbook 'bcpc', path: '../bcpc'
+cookbook 'bach_krb5', path: '../bach_krb5'
+cookbook 'smoke-tests', path: '../smoke-tests'
+#cookbook 'locking_resource', path: '../locking_resource'
```
This Berksfile needs some help, please chat with me tomorrow and I would be happy to offer advice. Please see [`cookbooks/bach_repository/Berksfile`](https://github.com/bloomberg/chef-bach/blob/master/cookbooks/bach_repository/Berksfile#L6-L7) for an example.

cbaenziger commented on cookbooks/bach_ambari/attributes/default.rb:
```diff
@@ -0,0 +1,45 @@
+# Ambari External Database attributes
+node.force_default['ambari']['db_type'] = 'mysql'
+
+mysql_port = node['bcpc']['hadoop']['mysql_port'] || 3306
```
We should have a simple wrapper recipe in BCPC/BCPC-Hadoop to populate BCPC attributes here.

cbaenziger commented on cookbooks/bach_ambari/metadata.rb:
```diff
@@ -0,0 +1,21 @@
+name 'bach_ambari'
+maintainer 'The Authors'
```
Again, please use BACH standards here.

cbaenziger commented on cookbooks/bcpc-hadoop/attributes/kerberos.rb:
```diff
@@ -127,7 +127,16 @@
     spnego_keytab: 'spnego.service.keytab'
-  }
+  },
+  ambari: {
```
Please see the [HDFS-DU work](https://github.com/bloomberg/chef-bach/pull/932/files#diff-dea32ef5d89946f76672ef3d6b565138) for how to include this using a wrapper recipe to make Ambari a component in BCPC-Hadoop rather than a base feature.

rashidaligee commented on cookbooks/ambari/README.md:
```diff
@@ -0,0 +1,61 @@
+============
+### Ambari Requirements
+
+To learn more about Ambari server and Ambari views, please see [latest documentation](https://docs.hortonworks.com/HDPDocuments/Ambari-2.6.1.3/bk_ambari-views/content/ch_understanding_ambari_views.html).
```
Hortonworks documentations are version based but I'll check with the team to see if there is a way to link documentation to latest/current version. Hortonworks documentation seems much better and relevant for user/administrator of ambari views, I'll double check to make sure I'm not missing if Apache is providing documentation with the same information.

pu239ppy replied:
How about https://github.com/apache/ambari

rashidaligee replied:
The intent is to land to HDP provided page as apache repo might have information, which HDP doesn't support .

amithkanand commented on cookbooks/ambari/recipes/ambari_views_setup.rb:
```diff
@@ -0,0 +1,40 @@
+  code <<-EOH
+    curl --user #{node['ambari']['admin']['user']}:#{node['ambari']['admin']['password']} -i -H 'X-Requested-By: ambari' -X POST #{node['ambari']['ambari_views_url']}/FILES/versions/1.0.0/instances/FILES_NEW_INSTANCE \
+--data '{
+  "ViewInstanceInfo" : {
```
Can we move this to an attribute? That way we add configuration for different views to the attribute and resource over here loops through the data structure and configures each view. Something similar to we do with [Kerberos](https://github.com/bloomberg/chef-bach/blob/master/cookbooks/bcpc-hadoop/attributes/kerberos.rb#L5-L22).

rashidaligee commented on cookbooks/ambari/recipes/ambari_views_setup.rb:
```diff
@@ -0,0 +1,40 @@
+bash 'create_instance_of_Files_view' do
+  # action :nothing
+  code <<-EOH
+    curl --user #{node['ambari']['admin']['user']}:#{node['ambari']['admin']['password']} -i -H 'X-Requested-By: ambari' -X POST #{node['ambari']['ambari_views_url']}/FILES/versions/1.0.0/instances/FILES_NEW_INSTANCE \
```
To explore http_request resource for this curl command.

amithkanand commented on cookbooks/ambari/templates/default/ambari.properties.erb:
```diff
@@ -0,0 +1,142 @@
+server.jdbc.database_name=<%= node['ambari']['databasename']%>
+server.jdbc.user.name=<%= node['ambari']['databaseusername']%>
+
+server.jdbc.user.passwd=/etc/ambari-server/conf/password.dat
```
Can we use `node['ambari']['ambari_server_conf_dir']` attribute here?

amithkanand commented on cookbooks/ambari/templates/default/ambari.properties.erb:
```diff
@@ -0,0 +1,142 @@
+server.jdbc.user.name=<%= node['ambari']['databaseusername']%>
+
+server.jdbc.user.passwd=/etc/ambari-server/conf/password.dat
+server.jdbc.rca.user.passwd=/etc/ambari-server/conf/password.dat
```
Same `node['ambari']['ambari_server_conf_dir']` over here as well.

cbaenziger commented on cookbooks/bach_ambari/attributes/default.rb:
```diff
@@ -0,0 +1,45 @@
+node.force_default['ambari']['install_java'] = false
+node.force_default['ambari']['java_home'] = '/usr/lib/jvm/java-8-oracle-amd64/'
```
This should come from [BCPC-Hadoop](https://github.com/bloomberg/chef-bach/blob/39fc7bbb30af803482308b08cc25955671bd5926/cookbooks/bcpc-hadoop/attributes/default.rb#L173)

cbaenziger replied:
Please see [bcpc-hadoop's default attributes file](https://github.com/bloomberg/chef-bach/blob/998a982e77aecad7949a6918145e9c2065d55629/cookbooks/bcpc-hadoop/attributes/default.rb#L172-L173)

cbaenziger commented on cookbooks/bach_ambari/attributes/default.rb:
```diff
@@ -0,0 +1,45 @@
+
+
+# FILES
+namenodes = get_namenodes.select{ |nn| nn[:bcpc][:node_number] }.map{ |nn| "namenode#{nn[:bcpc][:node_number]}" }.join(',')
```
Please do not call `get_namenodes`; the Chef node search code is a costly thing to execute. We have a function `set_hosts` which updates a number of node attributes to use instead. Please use the [node attributes](https://github.com/bloomberg/chef-bach/blob/e4352ff2604bf14b19cee62fffb045fc6e321e54/cookbooks/bcpc-hadoop/libraries/utils.rb#L186-L200) defined there.

rashidaligee commented:
Files, Hive, Tez and Workflow Manager views are integrated and tested in vm cluster.

cbaenziger commented on cookbooks/ambari/attributes/views.rb:
```diff
@@ -0,0 +1,49 @@
+# ambari Views Attributes
```
Why is this all commented?

rashidaligee replied:
This was to refactor code to use `http_request` resource. I guess, I'll delete the file and will get it from git history if needed.

cbaenziger commented on cookbooks/ambari/spec/unit/recipes/ambari_agent_setup_spec.rb:
```diff
@@ -0,0 +1,19 @@
+#
```
Why is this commented?

rashidaligee replied:
ambari-agent's are not required, it might be clean to remove the file.

cbaenziger commented on cookbooks/ambari/templates/default/ambari.properties.erb:
```diff
@@ -0,0 +1,142 @@
+#
```
Why is this commented?

cbaenziger commented on cookbooks/ambari/test/integration/default/serverspec/default_spec.rb:
```diff
@@ -0,0 +1,96 @@
+   its(:stdout) { should contain('Ambari Server running') }
+ end
+
+# it 'ambari server responds on port 8080' do
```
Still working on this?

rashidaligee commented:
Fixed issues found during deployment to lab cluster.
- extracted out ambari user.
- removed mysql-client package installation as it is available in Head nodes and Ambari Server will be installed in Head nodes.
- installed mysql-connector-java package as it is not installed by default in Head nodes.
- copy bach_ambari attributes to defaults recipe as float_host is dependent on bcpc-default recipe .

amithkanand commented:
This is tested successfully on a Lab cluster. One final change is required to fix the `apt` repo `key` issue.

rashidaligee commented on cookbooks/bach_ambari/recipes/default.rb:
```diff
@@ -0,0 +1,38 @@
+include_recipe 'bach_ambari::mysql_server_external_setup'
+
+# It is required to download ambari kerberos file.
+user "#{node['bcpc']['hadoop']['proxyuser']['ambari']}" do
```
Yes, please see `not_if` section.

rashidaligee commented on cookbooks/bcpc-hadoop/attributes/kerberos.rb:
```diff
@@ -127,6 +127,15 @@
+      principal: "#{node['bcpc']['hadoop']['proxyuser']['ambari']}",
+      keytab: 'ambari.service.keytab',
+      owner: "#{node['bcpc']['hadoop']['proxyuser']['ambari']}",
+      group: 'root',
```
Yes, in my editor, default settings are set to 2 spaces.

amithkanand commented on cookbooks/ambari/templates/default/ambari.properties.erb:
```diff
@@ -0,0 +1,142 @@
+server.jdbc.user.passwd=<%= File.join(node['ambari']['ambari_server_conf_dir'],'password.dat')%>
+server.jdbc.rca.user.passwd=<%= File.join(node['ambari']['ambari_server_conf_dir'],'password.dat')%>
+server.os_family=ubuntu14
+server.os_type=ubuntu14
```
Can we figure out the OS from Chef node object(`OHAI`)?

amithkanand commented on cookbooks/ambari/templates/default/ambari.properties.erb:
```diff
@@ -0,0 +1,142 @@
+server.execution.scheduler.maxDbConnections=5
+server.execution.scheduler.maxThreads=5
+server.execution.scheduler.misfire.toleration.minutes=480
+server.fqdn.service.url=http://169.254.169.254/latest/meta-data/public-hostname
```
What is this URL points to?

rashidaligee replied:
It turns out, `server.fqdn.service.url` property points to a server which provides fqdn of a host. It is generally used by ambari agent and would not affect our use case. Hortonworks community thread contains more information on it. https://community.hortonworks.com/questions/135132/where-is-serverfqdnserviceurl-hosted.html

cbaenziger replied:
For others' reference this is commonly the S3 EC2 meta-data service. E.g. on a [Chef-BCPC](https://github.com/bloomberg/chef-bcpc) OpenStack cluster, one would get the "floating IP" hostname rather than the "private IP"'s hostname from this REST call.

amithkanand commented on cookbooks/ambari/attributes/default.rb:
```diff
@@ -0,0 +1,104 @@
+# APT repository attributes
+default['apt']['compile_time_update'] = 'true'
+
+default['ambari']['repo_keyserver'] = 'keyserver.ubuntu.com'
```
Do we need this?

amithkanand commented on cookbooks/ambari/attributes/default.rb:
```diff
@@ -0,0 +1,104 @@
+default['ambari']['repo_keyserver'] = 'keyserver.ubuntu.com'
+default['ambari']['repo_key'] = 'hortonworks.key'#'B9733A7A07513CAD'
+node.default['ambari']['ambari_server_version'] = '2.6.1.5'
+node.default['ambari']['os']['ubuntu_version'] = 14
```
This can come from node object(using `OHAI`).

amithkanand commented on cookbooks/ambari/attributes/default.rb:
```diff
@@ -0,0 +1,104 @@
+node.default['ambari']['ambari_server_version'] = '2.6.1.5'
+node.default['ambari']['os']['ubuntu_version'] = 14
+
+node.default['ambari']['ambari_ubuntu_repo_url'] = "http://public-repo-1.hortonworks.com/ambari/ubuntu#{node['ambari']['os']['ubuntu_version']}/2.x/updates/#{node['ambari']['ambari_server_version']}"
```
We can get platform(`ubuntu`) from `OHAI`.
